### PR TITLE
[rubysrc2cpg] Enhancing TypeDecl implementation

### PR DESCRIPTION
**Pull request by khemrajrathore**
The PR contains the following
- For encountered TypeDecl create corresponding TypeRef assignments similar to how we do for python, so that types for the class instance usage within the same files is resolved
- For instance methods create a `this` parameter as the parameter with index 0, which help in tracking and linking dataflow for instance variables
- For method parameters refactored the current code to be short and precise along with adding corresponding `index` and `order` values
- Fixed 2 existing ignored test cases
cc - @fabsx00 @DavidBakerEffendi

### CODE REVIEW

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreatorHelper.scala [hunk 1]
```diff
@@ -36,6 +86,10 @@ trait AstCreatorHelper { this: AstCreator =>
     usedVariableNames.put(variableName, counter)
     currentVariableName
   }
+
+  protected def addReceiverEdge(dstNode: nodes.NewNode, srcNode: nodes.NewNode): Unit = {
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreatorHelper.scala:90]: Something like this exists in the `Ast()` class already

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala [hunk 2]
```diff
@@ -1314,9 +1317,25 @@ class AstCreator(
...
+    val astMethodName = astForMethodNamePartContext(ctx.methodNamePart())
+    val callNode      = astMethodName.head.nodes.filter(node => node.isInstanceOf[NewCall]).head.asInstanceOf[NewCall]
+
+    // Create thisParameter if this is an instance method
+    // TODO may need to revisit to make this more robust
+    val astMethodParamSeq = ctx.methodNamePart() match {
+      case _: SimpleMethodNamePartContext if !classStack.top.endsWith(":program") =>
+        val thisParameterNode = createMethodParameterIn(
+          "this",
+          typeFullName = callNode.methodFullName,
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala:1329]: Does this call node refer to the `:program` method?

**Review by DavidBakerEffendi — CHANGES_REQUESTED** [review]

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreatorHelper.scala (hunk 1 above)
Comment (khemrajrathore) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreatorHelper.scala:90]: Will remove this code as anyways we are not using this function anywhere in ruby

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala (hunk 2 above)
Comment (khemrajrathore) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala:1352]: No, the call node refers to methodDefinition. We use the information from the callNode and discard it

Basically here for any methodDefinition encountered we are categorizing the whether they are class method or instance method

**Comment by khemrajrathore** [comment]
@DavidBakerEffendi Have made the requested changes

**Review by DavidBakerEffendi — APPROVED** [review]

**Pull request merged**